wfe/wfe.go

```diff
@@ -1392,6 +1409,82 @@ func (wfe *WebFrontEndImpl) setCORSHeaders(response http.ResponseWriter, request
 	response.Header().Set("Access-Control-Max-Age", "86400")
 }
 
+// KeyRollover allows a user to change their signing key
+func (wfe *WebFrontEndImpl) KeyRollover(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
+	body, _, reg, prob := wfe.verifyPOST(ctx, logEvent, request, true, core.ResourceKeyChange)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Parse as JWS
+	newKey, parsedJWS, err := wfe.extractJWSKey(string(body))
+	if err != nil {
+		logEvent.AddError(err.Error())
+		wfe.sendError(response, logEvent, probs.Malformed(err.Error()), err)
+		return
+	}
+	payload, err := parsedJWS.Verify(newKey)
+	if err != nil {
+		logEvent.AddError("verification of the inner JWS with the inner JWK failed: %v", err)
+		wfe.sendError(response, logEvent, probs.Malformed("JWS verification error"), err)
+		return
+	}
+	var rolloverRequest struct {
+		NewKey  jose.JsonWebKey
+		Account string
+	}
+	err = json.Unmarshal(payload, &rolloverRequest)
+	if err != nil {
+		logEvent.AddError("unable to JSON parse resource from JWS payload: %s", err)
+		wfe.sendError(response, logEvent, probs.Malformed("Request payload did not parse as JSON"), nil)
+		return
+	}
+
+	if wfe.relativeEndpoint(request, fmt.Sprintf("%s%d", regPath, reg.ID)) != rolloverRequest.Account {
+		logEvent.AddError("incorrect account URL provided")
+		wfe.sendError(response, logEvent, probs.Malformed("Incorrect account URL provided in payload"), nil)
+		return
+	}
+
+	pnk, err := x509.MarshalPKIXPublicKey(rolloverRequest.NewKey.Key)
```

Comment:
Factoring out these three stanzas into a single `PublicKeyEqual` function would shorten this function a bit, and would make it more straightforward to specifically test these stanzas.